src/pgrn-custom-scan.c

```diff
@@ -181,19 +183,59 @@ PGrnBeginCustomScan(CustomScanState *customScanState,
 					EState *estate,
 					int eflags)
 {
-	grn_obj *sourcesTable;
+	PGrnScanState *state = (PGrnScanState *) customScanState;
+	TupleDesc tupdesc =
+		RelationGetDescr(customScanState->ss.ss_currentRelation);
 	Relation index = PGrnChooseIndex(customScanState->ss.ss_currentRelation);
+	grn_obj *sourcesTable;
+
 	if (!index)
 		return;
 	sourcesTable = PGrnLookupSourcesTable(index, ERROR);
-	elog(LOG, "DEBUG: custom-scan: %s", PGrnInspect(sourcesTable));
 	RelationClose(index);
+
+	GRN_PTR_INIT(&(state->columns), GRN_OBJ_VECTOR, GRN_ID_NIL);
+	for (unsigned int i = 0; i < tupdesc->natts; i++)
+	{
+		Form_pg_attribute attr = TupleDescAttr(tupdesc, i);
+		grn_obj *column =
+			PGrnLookupColumn(sourcesTable, NameStr(attr->attname), ERROR);
+		GRN_PTR_PUT(ctx, &(state->columns), column);
+	}
+	state->tableCursor = grn_table_cursor_open(
+		ctx, sourcesTable, NULL, 0, NULL, 0, 0, -1, GRN_CURSOR_ASCENDING);
 }
 
 static TupleTableSlot *
-PGrnExecCustomScan(CustomScanState *node)
+PGrnExecCustomScan(CustomScanState *customScanState)
 {
-	return NULL;
+	PGrnScanState *state = (PGrnScanState *) customScanState;
+	grn_id id;
+
+	if (!state->tableCursor)
+		return NULL;
+
+	id = grn_table_cursor_next(ctx, state->tableCursor);
+	if (!id)
+		return NULL;
+
+	{
+		TupleTableSlot *slot = customScanState->ss.ps.ps_ResultTupleSlot;
+		grn_obj value;
+
+		ExecClearTuple(slot);
+		GRN_VOID_INIT(&value);
```

Comment:
Could you move this to `PGrnScanState::column_value` or something and reuse this to avoid allocations as much as possible?